### ISSUE TITLE
How does eye movement work in coneAbsorptions.m?

### ISSUE THREAD

**Issue by xnieamo** [isetbio/isetbio#49]
I've read through the coneAbsorptions code.  It seems like the sensor is padded to account for the furthest possible location that the eye movement will bring it.  Then, the LMS cone absorptions are calculated for all the cones in this enlarged cone mosaic.  Afterwards, the original cone mosaic and the eye movement positions are used to get isomerizations due to eye movement.

Scaling the tremor amplitude by a large factor causes the movie to appear different from what I expected.

Here is some code:
```
fov = 2;
params.freq = 6; params.contrast = 1;
params.ph  = 0;  params.ang = 0;
params.row = 256; params.col = 256;
params.GaborFlag = 0.2; % standard deviation of the Gaussian window

% Set up scene and oi
scene = sceneCreate('harmonic', params);
scene = sceneSet(scene, 'h fov', fov);
oi = oiCreate('human');
oi = oiCompute(scene, oi);

% Create the eye movement object
em = emCreate;

sensor = sensorCreate('human');
sensor = sensorSetSizeToFOV(sensor, fov, scene, oi);
sensor = sensorSet(sensor, 'eyemove', em);
sensor = sensorSet(sensor, 'positions', zeros(500, 2));  % This is for 500 ms
sensor = sensorSet(sensor, 'exp time', 0.001);
sensor = sensorSet(sensor, 'time interval', 0.001);

% Magnify the amplitude to get a better visualization later
amp = emGet(em,'tremor amplitude');
em  = emSet(em,'tremor amplitude',100*amp);

sensor = sensorSet(sensor,'eye movement',em);
sensor = emGenSequence(sensor);

sensor = coneAbsorptions(sensor, oi);
photons = sensorGet(sensor, 'photons');
implay(photons/max(photons(:)));
```
The resulting movie looks like the image is floating around the window.  Looking at the coneAbsorptions.m code, I would have thought it should look like the eye panning around the scene.
Where would I be misunderstanding the code?

Xiaomao

**Comment by hjiang36** [comment]
Sorry for the late reply. I'm back now.

I ran the code posted above and the result looks somehow reasonable to me.

When tremor amplitude get scaled up with a large factor, eye movement get dominated by tremor. And tremor consists a series of independent Gaussian moves, which becomes a Brownian motion. If we plot the moving path, it's fairly close to Brownian motion.

Let me know if this makes sense.

**Comment by xnieamo** [comment]
The path of the eye movements make sense to me.  

I guess what was confusing me was how sensorCompute/coneAbsorptions deals with optical images and sensors that are different in size.  So if I have an OI of visual angle 1° and a sensor the size of 4°, will sensorCompute just set the extra area in the sensor to 0?  

Similarly, in the reverse case, when the OI is larger, I assume it just crops the appropriate section from the OI?

**Comment by wandell** [comment]
This is a very good question.  In starting to send you an answer, I realized that there are two things that are either bad or wrong in the code.

1.  The comments in oiPad said (until now) that the pad value around the image is 0.  In fact, the code pads the surround with the mean value of the oi photons.  We should probably set a flag that allows both options.  But over the years I had better luck with padding with the mean.  The error is that the comment at the top of oiPad said we pad with 0s.  I believe I fixed that.

2.  The sensor compute behavior is just what you infer.  It takes all the light there is in the oi.  If the sensor is bigger than the oi, then outside the 'pad' area the image is treated as 0s.

3.  Your comment led me to investigate one more thing.  The oi has a parameter for the horizontal field of view.  This is computed as the size in mm of the oi for one distance and the distance from the lens as another distance (image distance in mm).  It is bad that we have the parameter (h fov) and it can also be computed from other parameters (oi size) and (optics image distance).

If I had more time to act and check and so forth, I think I would get rid of the oiSet(oi,'h fov'), and I would replace the oiGet(oi,'h fov') with the calculation in oiPad that derives the horizontal field of view from the size and optics image distance.

**Comment by wandell** [comment]
Also, here is some code that shows the behavior when the oi field of view is smaller and larger than the sensor.

ieInit   % Apparently, we need to always run this at the start of any session.

% Make an MCC
scene = sceneCreate;
scene = sceneSet(scene,'fov',3);
oi = oiCreate('human');
oi = oiCompute(oi,scene);

% See the oi padding at the mean
vcAddObject(oi); oiWindow;

% Make a sensor
sensor = sensorCreate;
sensor = sensorSet(sensor,'size',[256,384]);
sensor = sensorCompute(sensor,oi);
sensor = sensorSet(sensor,'name',sprintf('FOV %.1f\n',sceneGet(scene,'fov')));
vcAddObject(sensor);

% Shrink the scene and do it again
scene = sceneSet(scene,'fov',0.5);
oi = oiCompute(oi,scene);
sensor = sensorCompute(sensor,oi);
sensor = sensorSet(sensor,'name',sprintf('FOV %.1f\n',sceneGet(scene,'fov')));

% Flip between the two sensors to see the padding reflected in the sensor image.
% Notice that the very small oi has a lot of black beyond the padding.
vcAddObject(sensor); sensorWindow;

**Comment by DavidBrainard** [comment]
I am trying to follow this.

The case I think I am most interested in understanding is when we define a sensor whose fov is small compared to the scene/oi.  In that case, I think the behavior we want is that the sensor stays the same time, and samples different regions of the oi according to the eye movements path that is set up.  So in other words, if I'm modeling the responses of a small set of foveal cones, what I get is what that fixed set of cones would see as the eye moves around.

I think the above is saying that that is what happens, but I'm not sure.  

Also, the implementation described seems a bit odd.  First the sensor is enlarged, then a full set of cone responses is computed, and then these are cropped down again.  But I don't see how that is going to work right if (as is the case for us) the mosaic packing isn't regular. 

Basically, I'm finding this all a bit confusing.

**Comment by xnieamo** [comment]
When the sensor is enlarged, the cone responses are calculated as if the sensor was entirely composed only of L, M, or S cones, not a mix of the three.  This gives us 3 isomerization matrices, one for each cone type.  Then, when it is cropped down, the mosaic packing of the original sensor is used to pick what values go into the final isomerization data, so I believe the sensor should stay the same.

**Comment by DavidBrainard** [comment]
> On Jul 21, 2015, at 1:17 PM, Xiaomao Ding <notifications@github.com> wrote:
> 
> When the sensor is enlarged, the cone responses are calculated as if the sensor was entirely composed only of L, M, or S cones, not a mix of the three. This gives us 3 isomerization matrices, one for each cone type. Then, when it is cropped down, the mosaic packing of the original sensor is used to pick what values go into the final isomerization data, so I believe the sensor should stay the same.
> 
Ah.

OK.

That seems like it will give the right answer, at least if we never get into mosaics where we model changes in outer segment length, inner segment diameter, and cone density with eccentricity.

Whether it is efficient or inefficient depends, I imagine, on how many eye positions are being sampled.

**Comment by xnieamo** [comment]
This seems resolved, so I'm closing the issue.
